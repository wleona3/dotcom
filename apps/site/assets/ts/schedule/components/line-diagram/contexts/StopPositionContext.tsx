import React, {
  createContext,
  MutableRefObject,
  PropsWithChildren,
  useCallback,
  useContext,
  useLayoutEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { LineDiagramStop } from "../../__schedule";
import {
  BASE_LINE_WIDTH,
  BRANCH_SPACING,
  StopCoord
} from "../graphics/graphic-helpers";
import { isMergeStop } from "../line-diagram-helpers";

type refType = HTMLElement | null;
export interface StopCoordState {
  [k: string]: StopCoord | null;
}
interface ContextValues {
  stopCoordState: StopCoordState;
  resetAllCoordinates: () => void;
  stopRefMap: Map<string, refType>;
}

/**
 * Computes pixel coordinates for a line diagram stop bubble
 * May be null if the stop is hidden on a collapsed branch
 */
function calcCoordinates(
  el: HTMLElement,
  stop: LineDiagramStop
): StopCoord | null {
  // x usually just a function of stop_data length
  const x = isMergeStop(stop)
    ? BASE_LINE_WIDTH + 1
    : BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
  // y should align with the ref position
  const { offsetTop, offsetHeight } = el;
  const y = offsetTop + offsetHeight / 2;
  return x && y ? [x, y] : null;
}

/**
 * Sets up a ref for every stop,
 * sets up state to hold coordinate values,
 * sets up a function to update coordinates based on the refs.
 */
const useStopPositions = (stops: LineDiagramStop[]): ContextValues => {
  const map = new Map(
    stops.map<[string, refType]>(stop => [stop.route_stop.id, null])
  );
  const objEntries = stops.map<[string, StopCoord | null]>(stop => [
    stop.route_stop.id,
    null
  ]);
  const stopRefMap: MutableRefObject<Map<string, refType>> = useRef(map);
  const initialObject = Object.fromEntries(objEntries);
  const [stopCoordState, setStopCoordState] = useState<StopCoordState>(
    initialObject
  );

  const resetAllCoordinates = useCallback((): void => {
    window.requestAnimationFrame((): void => {
      // get new coordinates
      setStopCoordState(
        (state): StopCoordState => {
          const newState = { ...state };
          stops.forEach(stop => {
            const stopId = stop.route_stop.id;
            const el = stopRefMap.current.get(stopId);
            newState[stopId] = el ? calcCoordinates(el, stop) : null;
          });
          return newState;
        }
      );
    });
  }, [stops, stopRefMap]);

  // only update when the state actually changes (not every render)
  const contextValues = useMemo(
    () => ({
      stopCoordState,
      resetAllCoordinates,
      stopRefMap: stopRefMap.current
    }),
    [stopCoordState, resetAllCoordinates]
  );

  // Automatically update positions after resize events
  useLayoutEffect(() => {
    window.addEventListener("resize", resetAllCoordinates);
    return () => window.removeEventListener("resize", resetAllCoordinates);
  }, [resetAllCoordinates]);

  return contextValues;
};

/**
 * Sets up context that will hold:
 * stopCoordState: an object containing updated values for pixel location
 * resetAllCoordinates: a function to recalculate coordinates for every stop
 */
export const StopPositionContext = createContext<ContextValues>({
  stopCoordState: {},
  resetAllCoordinates: () => {},
  stopRefMap: new Map()
});

// Get latest coordinates for one or all stops
export function useStopPositionCoordinates(): StopCoordState | null;
export function useStopPositionCoordinates(stopId?: string): StopCoord;
export function useStopPositionCoordinates(
  stopId?: string
): StopCoordState | StopCoord | null {
  const { stopCoordState } = useContext(StopPositionContext);
  if (stopId) {
    return stopCoordState[stopId];
  }
  return stopCoordState;
}

// Triggers recalculation of all coordinates
export function useStopPositionReset(): () => void {
  const { resetAllCoordinates } = useContext(StopPositionContext);
  return resetAllCoordinates;
}

// Returns the set of refs being used to track <h4> stop names
export const useStopRefs = (): Map<string, refType> => {
  const { stopRefMap } = useContext(StopPositionContext);
  return stopRefMap;
};

/**
 * Access context values within child components via
 *  useStopPositionCoordinates(stopId?)
 *  useStopPositionReset()
 *  useStopPositionRefs()
 */
export const StopPositionProvider = ({
  stops,
  children
}: PropsWithChildren<{ stops: LineDiagramStop[] }>): React.ReactElement => {
  const contextValues = useStopPositions(stops);
  return (
    <StopPositionContext.Provider value={contextValues}>
      {children}
    </StopPositionContext.Provider>
  );
};
