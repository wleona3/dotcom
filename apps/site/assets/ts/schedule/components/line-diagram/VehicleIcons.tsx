import React from "react";
import { renderToString } from "react-dom/server";
import { RouteStop, LineDiagramVehicle } from "../__schedule";
import CrowdingPill from "./CrowdingPill";
import { TooltipWrapper, vehicleArrowIcon } from "../../../helpers/icon";
import { CIRC_RADIUS } from "./graphics/graphic-helpers";
import { useStopPositionCoordinates } from "./contexts/StopPositionContext";

interface VehicleIconsProps {
  stop: RouteStop;
  vehicles: LineDiagramVehicle[] | null;
}

const CrowdingIconString = (vehicle: LineDiagramVehicle): string =>
  renderToString(<CrowdingPill crowding={vehicle.crowding} />);

const VehicleIcons = ({
  stop,
  vehicles
}: VehicleIconsProps): JSX.Element | null => {
  const coords = useStopPositionCoordinates(stop.id);
  if (!vehicles || !coords) return null;
  const [x, y] = coords;
  const left = `${x - CIRC_RADIUS}px`;
  const tooltips = vehicles.map(vehicle => {
    const top = `${{
      // eslint-disable-next-line camelcase
      in_transit: y - 50,
      incoming: y - 25,
      stopped: y - 10
    }[vehicle.status] || y}px`;

    return (
      <div
        key={vehicle.id}
        className="m-schedule-diagram__vehicle"
        style={{ top, left }}
      >
        <TooltipWrapper
          tooltipText={`<div class="m-schedule-diagram__vehicle-tooltip">${
            vehicle.crowding ? `${CrowdingIconString(vehicle)}<br/>` : ""
          }${vehicle.tooltip}</div>`}
          tooltipOptions={{ placement: "right", animation: false, html: true }}
        >
          {vehicleArrowIcon("m-schedule-diagram__vehicle--icon")}
        </TooltipWrapper>
      </div>
    );
  });

  return <>{tooltips}</>;
};

export default VehicleIcons;
