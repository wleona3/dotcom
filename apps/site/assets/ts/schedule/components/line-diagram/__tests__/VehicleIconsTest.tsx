import React from "react";
import {
  LineDiagramVehicle,
  RouteStop,
  LineDiagramStop
} from "../../__schedule";
import VehicleIcons from "../VehicleIcons";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";
import { StopCoordState } from "../contexts/StopPositionContext";

const mockState: StopCoordState = {
  "test-stop": [12, 80]
};
const stop = { id: "test-stop", name: "Test Stop" } as RouteStop;
const lineDiagramStop: LineDiagramStop = {
  stop_data: [],
  route_stop: stop,
  alerts: []
};
const vehicles = [
  { id: "vehicle-1", status: "in_transit", tooltip: "vehicle 1 tooltip text" },
  { id: "vehicle-2", status: "incoming", tooltip: "vehicle 2 tooltip text" },
  { id: "vehicle-3", status: "stopped", tooltip: "vehicle 3 tooltip text" }
] as LineDiagramVehicle[];

describe("VehicleIcons with no vehicles", () => {
  it("doesn't render", () => {
    const { container } = mockedRenderWithStopPositionContext(
      <VehicleIcons stop={stop} vehicles={null} />,
      [lineDiagramStop],
      mockState
    );

    expect(container.innerHTML).toBeFalsy();
  });
});

describe("VehicleIcons with vehicles", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeAll(() => {
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <VehicleIcons stop={stop} vehicles={vehicles} />,
      [lineDiagramStop],
      mockState
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders each vehicle", () => {
    const vehicleNodes = container.querySelectorAll(
      ".m-schedule-diagram__vehicle"
    );
    expect(vehicleNodes).toHaveLength(vehicles.length);
  });

  it.each`
    index | tooltip
    ${0}  | ${"vehicle 1 tooltip text"}
    ${1}  | ${"vehicle 2 tooltip text"}
    ${2}  | ${"vehicle 3 tooltip text"}
  `("positions vehicles according to status $status", ({ index, tooltip }) => {
    const node = container.querySelectorAll(".m-schedule-diagram__vehicle")[
      index
    ];
    expect(node.querySelector("[tooltipText]")?.textContent).toContain(tooltip);
  });

  it.each`
    index | expectedPosition | status
    ${0}  | ${30}            | ${"in_transit"}
    ${1}  | ${55}            | ${"incoming"}
    ${2}  | ${70}            | ${"stopped"}
  `(
    "positions vehicles according to status $status",
    ({ index, expectedPosition }) => {
      const node = container.querySelectorAll<HTMLElement>(
        ".m-schedule-diagram__vehicle"
      )[index];
      const { top } = node.style; // e.g. "30px"
      const top_number = parseInt(top.substring(0, 2)); // e.g. 30
      expect(top_number).toEqual(expectedPosition);
    }
  );
});

it("VehicleIcons includes the vehicle crowding status if available", () => {
  const { container } = mockedRenderWithStopPositionContext(
    <VehicleIcons
      stop={stop}
      vehicles={[
        {
          id: "v1",
          status: "incoming",
          crowding: "some_crowding",
          tooltip: "tooltip text"
        }
      ]}
    />,
    [lineDiagramStop],
    mockState
  );
  const node = container.querySelector(".m-schedule-diagram__vehicle");
  expect(node!.querySelector("[tooltipText]")?.textContent).toContain(
    "Some crowding"
  );
});
