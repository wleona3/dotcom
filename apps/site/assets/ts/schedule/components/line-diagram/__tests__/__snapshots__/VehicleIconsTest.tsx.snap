// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleIcons with vehicles renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-diagram__vehicle"
    style="top: 30px; left: 8px;"
  >
    <span
      data-animation="false"
      data-html="true"
      data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 1 tooltip text</div>"
      data-placement="right"
      data-selector="true"
      data-toggle="tooltip"
      data-trigger="hover focus"
    >
      <span
        aria-hidden="false"
        class="notranslate m-schedule-diagram__vehicle--icon"
      >
        SVG
      </span>
    </span>
  </div>
  <div
    class="m-schedule-diagram__vehicle"
    style="top: 55px; left: 8px;"
  >
    <span
      data-animation="false"
      data-html="true"
      data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 2 tooltip text</div>"
      data-placement="right"
      data-selector="true"
      data-toggle="tooltip"
      data-trigger="hover focus"
    >
      <span
        aria-hidden="false"
        class="notranslate m-schedule-diagram__vehicle--icon"
      >
        SVG
      </span>
    </span>
  </div>
  <div
    class="m-schedule-diagram__vehicle"
    style="top: 70px; left: 8px;"
  >
    <span
      data-animation="false"
      data-html="true"
      data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 3 tooltip text</div>"
      data-placement="right"
      data-selector="true"
      data-toggle="tooltip"
      data-trigger="hover focus"
    >
      <span
        aria-hidden="false"
        class="notranslate m-schedule-diagram__vehicle--icon"
      >
        SVG
      </span>
    </span>
  </div>
</DocumentFragment>
`;
